### ISSUE TITLE
redirect from getInitialProps ?

### ISSUE THREAD

**Issue by matthewmueller** [zeit/next.js#649]
Noticing that if I try to redirect inside getInitialProps, it'll still render the component (with no props)

Here's an example:

```js
export default async function auth (ctx) {
  const { token } = cookies(ctx)
  if (ctx.req && !token) {
    ctx.res.writeHead(302, { Location: '/' })
    return
  } else if (!token) {
    document.location.pathname = '/'
    return
  }

  const res = await loadUser(token)

  if (res instanceof Error) {
    return ctx.res.writeHead(302, { Location: '/' })
  }

  return res.user
}
```

Wondering what I can do in the:

```js
if (ctx.req && !token) {
  ctx.res.writeHead(302, { Location: '/' })
  return
} else if (!token) {
  document.location.pathname = '/'
  return
}
```

To stop from rendering the component and just redirect.

**Comment by nkzawa** [comment]
It should work when you call `res.end()` as well.

**Comment by PetrSnobelt** [comment]
It works with `res.end()` 
👍

**Comment by thonatos** [comment]
express.js 4.x

```
class Index extends Component {

  static getInitialProps = (ctx) => {
    const { req, res, query } = ctx    
    if(req.isMobile){
      res.redirect('/summer-travel-mobile')
    }
    return {}
  }
  render(){
      return (<div>temp page</div>)
  }
}
```

😀

**Comment by slorenzo** [comment]
I have used the following way to redirect depend of the context:

```
static getInitialProps = (ctx) => {
    const { isServer, res } = ctx

    if(isServer){
      res.redirect('/dashboard')
      res.end()
    }

    if(!isServer) {
      Router.push('/dashboard')
    }

    return {}
}
```

**Comment by eyaleizenberg** [comment]
@oliviertassinari @slorenzo  where do you get the Router object from?

**Comment by mario-jerkovic** [comment]
@eyaleizenberg its imported as `import Router from 'next/router'`

**Comment by kaimiyang** [comment]
i use the follow code 
`res.end(`<script>window.location.href = ${path.login}</script>`)`

**Comment by KumarAbhirup** [comment]
Does not work with Next v9.
There it says, res.redirect is not a function.

**Comment by juancamiloqhz** [comment]
I've found this post with a solution to the "res.redirect is not a function"
`Home.getInitialProps = ({ res }) => {

  if (res) {
    res.writeHead(301, {
      Location: 'new/url/destination/here'
    });
    res.end();
  }

  return {};
};`

Can someone confirm this solution?

**Comment by valmack** [comment]
@juancamiloqhz 
> Can someone confirm this solution?

Yes this is what I am doing for my server side redirects and it works.

My solution looks like this 

```js
// paper/index.js

import Router from "next/router";
import { redirect } from "~/config/utils";

const redirectPath = "summary";

function Paper() {
  Router.push(redirectPath);
}

Paper.getInitialProps = async (ctx) => {
  redirect(ctx, redirectPath);
  return {};
};

export default Paper;
```

```js
// utils.js

export function redirect(ctx, path) {
  const { res } = ctx;
  if (res) {
    res.writeHead(301, { Location: path });
    res.end();
  }
}
```

**Comment by jmaister** [comment]
@valmack Just as a remainder for HTTP codes:

301 Moved permanently (The URL will not exist anymore and this is the new one)
302 Found (The URL works but you have to go to this one first)

**Comment by OlegBrony** [comment]
> Paper.getInitialProps = async (ctx) => {
>   redirect(ctx, redirectPath);
>   return {};
> };
> 
> export default Paper;
> ```
> 
> ```js
> // utils.js
> 
> export function redirect(ctx, path) {
>   const { res } = ctx;
>   if (res) {
>     res.writeHead(301, { Location: path });
>     res.end();
>   }
> }
> ```

how to implement it with new getStaticProps funciton?

**Comment by TommySorensen** [comment]
> @valmack Just as a remainder for HTTP codes:
> 
> 301 Moved permanently (The URL will not exist anymore and this is the new one)
> 302 Found (The URL works but you have to go to this one first)

Or use:
```
307 Temporary Redirect
308 Permanent Redirect (is best to use, even better then 301)
```

**Comment by mngyuan** [comment]
@valmack, your code results in `Uncaught (in promise) Error: No router instance found. You should only use "next/router" inside the client side of your app.` for me, because the `Router` call is happening inside the render, which happens serverside on a cold page load. To get it to work and only use Router client side, I'm wrapping it in a `useEffect` call like so:
```js
// pages/work.js
import {useEffect} from 'react';
import Router from 'next/router';

const path = '/work/project-a';

function Work() {
  useEffect(() => {
    Router.push(path);
  });
  return <></>;
}

export async function getInitialProps(ctx) {
  const {res} = ctx;
  if (res) {
    res.writeHead(301, {Location: path});
    res.end();
  }
  return {};
}

export default Work;
```

**Comment by gfs980** [comment]
For those who still struggle with getInitialProps with redux.
I have made this solution and it is working for me perfectly fine.

`
const UseAuthentication = (ChildComponent) => {
    const PrivateRouteComponent = (dataStuff) => {
        console.log('dataStuff: ', dataStuff);

        return (
            <Private>
                <ChildComponent />
            </Private>
        )
    }

    PrivateRouteComponent.getInitialProps = ({ req, res, err}) => {
        const auth = store.getState().user.isAuthenticated;
        console.log('PrivateRouteComponent.getInitialProps auth: ', auth);

        if(!auth) {
            res.writeHead(301, { Location: '/login' });
            res.end();
        }

        return {
            props: {
                data: 'Blah Blah GEORGE'
            },
        }
    }


    return PrivateRouteComponent
}

export default UseAuthentication;
`
I did need to access my redux store to check if user authenticated or not. Basically it is some sort of custom hook wrapper for PrivateRoutes that should to be accessed only if the user is authenticated.

I am using Next.js 10.0.3 latest at the moment of this post :D

I was not able to do redux wrapper with 'next-redux-wrapper' since it wont work with i18n translations unfortunatelly :(

So here is my store file with redux-thunk:
`
import { composeWithDevTools } from 'redux-devtools-extension';
import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';

export interface ApplicationState {
  user: UserState,
  card: CardState,
  error: ErrorState,
  uploadFile:any,
  realty: RealtyState,
  message: MessageState
}

export const rootReducer: Reducer<ApplicationState> = combineReducers<ApplicationState>({
  user: userReducer,
  card: cardReducer,
  error: errorReducer,
  uploadFile,
  realty: realtyReducer,
  message: messageReducer
});

export const store = createStore(
    rootReducer,
    applyMiddleware(thunkMiddleware)
);
`

And here is my _app.jsx file:
`
import { store } from "../store/store";
import { appWithTranslation } from '../i18nNew'
import { Provider } from 'react-redux'

const MyApp = ({ Component, pageProps }) => {
  const router = useRouter();

  React.useEffect(() => {
      console.log('App rendering')
  }, [router]);

  return (
    <Provider store={store} >
      <Layout>
        <Component {...pageProps} />
      </Layout>
    </Provider>
  )
}

MyApp.getInitialProps = async (appContext) => ({ ...await App.getInitialProps(appContext) })

export default appWithTranslation(MyApp)
`

and finaly my i18n file: 
`
import NextI18next from "next-i18next"
import { useTranslation as originalUseTranslation } from "react-i18next"
const path = require('path')

export const nextI18next = new NextI18next({
    localePath: path.resolve('./public/static/locales'),
    defaultLanguage: 'ru',
    otherLanguages: ['en', 'de'],
})

export const appWithTranslation = nextI18next.appWithTranslation;
export const useTranslation = originalUseTranslation;

`
